Support ATN2 Transact-SQL function

LareinaWei opened the pull request:
Make use of PG's atan2() function to implement atn2 fuction.

Task: BABEL-725
Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

### Description

Currently, babelfish does not support atn2() Transact-SQL function. This commit implements atn2() using pg's atan2() function.

### Issues Resolved

Babelfish do not support ATN2 Transact-SQL fucntion.

### Reference Document Link
https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/t-sql/functions/atn2-transact-sql.md

### Test Scenarios Covered
* **Use case based -**
  Added a new test `functions/atn2`
* **Boundary conditions -**
  Covered in test `functions/atn2`
* **Arbitrary inputs -**
 added
* **Negative test cases -**
  Float overflow would thrown an error.
* **Minor version upgrade tests -**
  tests added
* **Major version upgrade tests -**
  tests added
* **Performance tests -**
  N/A
* **Tooling impact -**
  None
* **Client tests -**
  None

### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Support ATN2 Transact-SQL function

Make use of PG's atan2() function to implement atn2 fuction.

Task: BABEL-725
Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

2jungkook commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -987,6 +987,18 @@ END;
 $BODY$
 LANGUAGE plpgsql PARALLEL SAFE IMMUTABLE RETURNS NULL ON NULL INPUT;
 
+CREATE OR REPLACE FUNCTION sys.atn2(IN x DOUBLE PRECISION, IN y DOUBLE PRECISION) RETURNS DOUBLE PRECISION
```
Let's use T-SQL types if possible, e.g., sys.float.

2jungkook commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -987,6 +987,18 @@ END;
+    RETURN res;
+END;
+$BODY$
+LANGUAGE plpgsql PARALLEL SAFE IMMUTABLE RETURNS NULL ON NULL INPUT;
```
Can we use `SQL` instead of `plpgsql`?

LareinaWei replied:
It keeps throwing syntax error `~~ERROR (Message: syntax error at or near "SYS")~~` after I changed the language from `plpgsql` to `sql`... I tried change 1)nothing but the language, 2)every above suggested changes including changing the language, 3) one of above suggested change at a time along with changing the language, and they all throw syntax errors...

No error is thrown when not changing the language (but change everything else).

2jungkook replied:
This would work
```
CREATE OR REPLACE FUNCTION sys.atn2(IN x sys.float, IN y sys.float)
RETURNS sys.float
AS
$$
    SELECT pg_catalog.atan2(x, y);
$$
LANGUAGE SQL PARALLEL SAFE IMMUTABLE RETURNS NULL ON NULL INPUT;
```

2jungkook commented on test/JDBC/expected/BABEL-725-vu-cleanup.out:
```diff
@@ -0,0 +1,110 @@
+DROP VIEW BABEL_725_vu_prepare_v1;
```
Lately, we are trying to use more appropriate test file names. What about `functions/atn2-vu-XXX`?

2jungkook commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -987,6 +987,18 @@ END;
+DECLARE
+    res DOUBLE PRECISION;
+BEGIN
+    res = atan2(x, y);
```
When we use a PG's function, let's make it clear by adding `pg_catalog`, e.g., `pg_catalog.atan2`.

2jungkook commented on test/JDBC/expected/BABEL-725-vu-prepare.out:
```diff
@@ -0,0 +1,206 @@
+GO
+
+-- test with different datatype
+CREATE VIEW BABEL_725_vu_prepare_v2 AS (
```
We need to test all data types that can be implicitly casted to `float` (ref: https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/t-sql/data-types/data-type-conversion-database-engine.md)

2jungkook commented:
In the PR description, let's add the followings:
1. Reference document link: https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/t-sql/functions/atn2-transact-sql.md
2. Test Scenarios Covered section (ref: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/874)

Commit:
Renamed test files to include tested function name, added tests for
several datatypes to cover all datatypes that can be implicitly casted
to float.

Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

Commit:
Changed function language to SQL, used t-sql types for variables, stated called pg function by adding pg_catalog.

Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

Commit:
Removed covered dependency in
upgrade_validationa/expected_dependency.out

Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

Commit:
Added tests for const number and const literal, removed function tests.

Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

kuntalghosh commented on test/JDBC/input/functions/atn2-vu-prepare.sql:
```diff
@@ -0,0 +1,110 @@
+        ATN2('2.5', '3.5') AS res3
+    );
+GO
+
```
This doesn't have any test cases for boundary scenarios and NULL conditions. For example, ATN(0,0) throws an error in SQL Server, but PG shows the result as 0.

LareinaWei replied:
I'll add the test case for input (0,0) and modify the function to have the same behavior with SQL Server. And actually I have NULL test cases which are at ~ line 30.

Commit:
Modified function to throw error when input is (0,0),
added test cases for input (0,0).

Task: BABEL-725
Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

LareinaWei committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-725

Commit:
empty commit

Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

LareinaWei committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-725
